src/main/webapp/FacebookAPI.js

```diff
@@ -0,0 +1,28 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Handles calls to the Facebook API, provides wrapper methods with callbacks
+class FacebookAPI {
+  static getUserFacebookInfoAndRedirect(callback) {
+    return callback();
+  }
```

Comment:
Hmm IIUC these aren't actually doing anything; it'll just always execute callback() immediately, so nothing is being mocked out -- it'll still call through to FB.

Instead I was thinking about something like:

```js
class FacebookApi {
  static getUserFacebookInfoAndRedirect(callback) {
    FB.api('/me?fields=id,name,email,friends', callback);
  }
}
```

Then use it like:

```js
FacebookApi.getUserFacebookInfoAndRedirect(function(response) {
  let userInfo = new URLSearchParams({
    name: response.name,
    id: response.id,
    email: response.email,
  });

  // Store user information in Datastore
  // TODO: Add post request
  // Redirect to the profile page, and pass the user's info in the URL
  window.location.href = "./profile.html?" + userInfo.toString();
});
```

And in your test, you can override it to invoke the callback with a mock response, something like:

```
FacebookApi.getUserFacebookInfoAndRedirect = function(callback) {
  callback({ /* construct a mock response here */ });
}; 
```

(my syntax may not be correct, I'm not really familiar with OO in Javascript)

Reply:
The first two code snippets make sense, but the last is a little confusing. When testing this new function, what would I be inserting as the callback parameter? Also should the mock response be something similar to what I had before, like returning a string value?

Reply:
Also I've been a bit blocked on the fact that Jest doesn't 'work' (it errors immediately) when run on the code in `facebook-login.js` that loads the Facebook API async... so referencing the `FB` object in the `FacebookAPI.js` file doesn't pass any tests or compile the project. Not sure how to proceed here.

Reply:
The last snippet is overriding the FacebookApi.getUserFacebookInfoAndRedirect function. But actually, maybe it'd be better to follow the Jest documentation -- it looks like there is first-class support for mocking:

https://jestjs.io/docs/en/mock-functions

(happy to help you figure it out if it's still unclear from those docs)

Also if your test code is failing to load the API, I suspect you'll want to separate the code which loads the API from the code which actually uses it -- so that you can avoid loading it in the test and just use the mock.